resourcemanager/gpool/spool/spool.go

```diff
@@ -0,0 +1,163 @@
+	options            *Options
+	capacity           atomic.Int32
+	running            atomic.Int32
+	isStop             atomic.Bool
+	concurrencyMetrics prometheus.Gauge
+	taskManager        pooltask.TaskManager[any, any, any, any, pooltask.NilContext]
+	gpool.BasePool
+}
+
+// NewPool create a single goroutine pool.
+func NewPool(name string, size int32, component util.Component, options ...Option) (*Pool, error) {
+	opts := loadOptions(options...)
+	result := &Pool{
+		options:            opts,
+		concurrencyMetrics: metrics.PoolConcurrencyCounter.WithLabelValues(name),
+		taskManager:        pooltask.NewTaskManager[any, any, any, any, pooltask.NilContext](size), // TODO: this general type
+	}
+	result.SetName(name)
+	result.capacity.Store(size)
+	result.concurrencyMetrics.Set(float64(size))
+	err := resourcemanager.InstanceResourceManager.Register(result, name, component)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+// Tune changes the pool size.
+func (p *Pool) Tune(size int) {
+	p.SetLastTuneTs(time.Now())
+	p.capacity.Store(int32(size))
+	p.concurrencyMetrics.Set(float64(size))
+}
+
+// Run runs a function in the pool.
+func (p *Pool) Run(fn func()) error {
+	if p.isStop.Load() {
+		return gpool.ErrPoolClosed
+	}
+	_, run := p.isRunnable(1, true)
+	if !run {
+		return gpool.ErrPoolOverload
+	}
+	p.taskManager.RegisterTask(p.NewTaskID(), 1)
+	p.run(fn)
+	return nil
+}
+
+// Cap returns the capacity of the pool.
+func (p *Pool) Cap() int {
+	return int(p.capacity.Load())
+}
+
+// Running returns the number of running goroutines.
+func (p *Pool) Running() int {
+	return int(p.running.Load())
+}
+
+func (p *Pool) run(fn func()) {
+	p.wg.Add(1)
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logutil.BgLogger().Error("recover panic", zap.Any("recover", r))
+			}
+			p.wg.Done()
+			p.running.Add(-1)
+		}()
+		fn()
+	}()
+}
+
+// RunWithConcurrency runs a function in the pool with concurrency.
+func (p *Pool) RunWithConcurrency(fn func(), concurrency int) error {
+	if p.isStop.Load() {
+		return gpool.ErrPoolClosed
+	}
+	conc, run := p.isRunnable(int32(concurrency), false)
+	if !run {
+		return gpool.ErrPoolOverload
+	}
+	// TODO: taskManager need to refactor
+	p.taskManager.RegisterTask(p.NewTaskID(), conc)
+	for n := int32(0); n < conc; n++ {
+		p.run(fn)
+	}
+	return nil
+}
+
+// isRunnable is to check if the pool is runnable.
+func (p *Pool) isRunnable(concurrency int32, mustEnough bool) (conc int32, run bool) {
+	for {
+		value, run := p.isRunnableInternal(concurrency, mustEnough)
+		if run {
+			return value, run
+		}
+		if p.options.Nonblocking {
+			return 0, false
+		}
+		time.Sleep(5 * time.Millisecond)
+	}
+}
+
+func (p *Pool) isRunnableInternal(concurrency int32, mustEnough bool) (conc int32, run bool) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if p.running.Load() >= p.capacity.Load() {
+		return 0, false
+	}
+	if p.running.Load()+concurrency < p.capacity.Load() {
```

Comment:
Should it be `<=`? I considered the following situation:

If `mustEnough` is `true`, the `concurrency` is always 1 (as shown in the code). Then the `p.capacity.Load() - p.running.Load() > 0` and `p.running.Load()+1 < p.capacity.Load()` could be unified if this is an `<=`, because `p.capacity.Load() - p.running.Load() > 0` iff `p.running.Load()+1 <= p.capacity.Load()`

Reply:
Done